Source/Application/Use Cases/Restaurant List/Controller/RestaurantListController.swift

```diff
@@ -0,0 +1,73 @@
+import UIKit
+
+final class RestaurantListController: NSObject {
+
+    private static let listCellReuseIdentifier = "_restaurant_list_cell"
+    private let dependencies: HasImageProvider
+
+    private let restaurants: [Restaurant]
+
+    private var images: [Restaurant: UIImage] = Dictionary()
+    private var isLoadingImage: [Restaurant: Bool] = Dictionary()
+
+    init(dependencies: HasImageProvider, restaurants: [Restaurant]) {
+        self.dependencies = dependencies
+        self.restaurants = restaurants
+    }
+
+    func prepare(tableView: UITableView) {
+        tableView.register(RestaurantListCell.self, forCellReuseIdentifier: RestaurantListController.listCellReuseIdentifier)
+    }
+}
+
+extension RestaurantListController: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return self.restaurants.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: RestaurantListController.listCellReuseIdentifier, for: indexPath)
+        if let cell = cell as? RestaurantListCell {
+
+            let restaurant = self.restaurants[indexPath.row]
+
+            cell.update(with: RestaurantListCellViewModel(restaurant: restaurant))
+
+            if let image = self.images[restaurant] {
+                cell.updateImage(image)
+            } else {
+                switch self.isLoadingImage[restaurant] {
```

Comment:
Here's another piece of state that I think we may be able to better encapsulate, as here, though it's directly related to whether or not there is, or will be, an image inside the dictionary, it is only loosely bound to the state of the dictionary and is imperatively updated.

What do you think about the situation?